src/Interpreters/InstrumentationManager.cpp

```diff
@@ -0,0 +1,530 @@
+    }
+
+    std::lock_guard lock(shared_mutex);
+    auto it = instrumented_points.get<InstrumentedPointKey>().find(InstrumentedPointKey{function_id, entry_type, handler_name_lower});
+    if (it != instrumented_points.get<InstrumentedPointKey>().end())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Handler of this type is already installed for function_id '{}', function name '{}'",
+            function_id, function_name);
+    }
+
+    patchFunctionIfNeeded(function_id);
+
+    InstrumentedPointInfo info{context, instrumented_point_ids, function_id, function_name, handler_name_lower, entry_type, parameters};
+    LOG_DEBUG(logger, "Adding instrumentation point for {}", info.toString());
+    instrumented_points.emplace(std::move(info));
+    instrumented_point_ids++;
+}
+
+void InstrumentationManager::unpatchFunction(std::variant<UInt64, bool> id)
+{
+    std::lock_guard lock(shared_mutex);
+
+    if (std::holds_alternative<bool>(id))
+    {
+        LOG_DEBUG(logger, "Removing all instrumented functions");
+        for (const auto & info : instrumented_points)
+            unpatchFunctionIfNeeded(info.function_id);
+        instrumented_points.clear();
+    }
+    else
+    {
+        std::optional<InstrumentedPointKey> function_key;
+        InstrumentedPointInfo instrumented_info;
+        for (const auto & info : instrumented_points)
+        {
+            if (info.id == std::get<UInt64>(id))
+            {
+                function_key = InstrumentedPointKeyExtractor()(info);
+                instrumented_info = info;
+                break;
+            }
+        }
+
+        if (!function_key.has_value())
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Unknown instrumentation point id to remove: ({})", std::get<UInt64>(id));
+
+        LOG_DEBUG(logger, "Removing instrumented function {}", instrumented_info.toString());
+        unpatchFunctionIfNeeded(function_key->function_id);
+        instrumented_points.get<InstrumentedPointKey>().erase(function_key.value());
+    }
+}
+
+InstrumentationManager::InstrumentedPoints InstrumentationManager::getInstrumentedPoints()
+{
+    SharedLockGuard lock(shared_mutex);
+    InstrumentedPoints points;
+    points.reserve(instrumented_points.size());
+
+    for (const auto & info : instrumented_points)
+        points.emplace_back(info);
+
+    return points;
+}
+
+void InstrumentationManager::dispatchHandler(Int32 func_id, XRayEntryType entry_type)
+{
+    InstrumentationManager::instance().dispatchHandlerImpl(func_id, entry_type);
+}
+
+void InstrumentationManager::dispatchHandlerImpl(Int32 func_id, XRayEntryType entry_type)
+{
+    /// We don't need to distinguish between a normal EXIT and a TAIL EXIT, so we convert
+    /// the latter to the former to simplify the rest of the logic.
+    if (entry_type == XRayEntryType::TAIL)
+        entry_type = XRayEntryType::EXIT;
+
+    for (const auto & [handler_name, handler_function] : handler_name_to_function)
```

Comment:
This will always trigger all 3 handlers (and potentially more) to lock mutex and search the same data structure. It would be better to invert the logic: find the tracepoint corresponding to the func_id once, then iterate through all attached handlers if any.

Just my imagination: this will also unlock the possibility of registering multiple handlers of the same type. This may be is pointless with LOG/SLEEP/PROFILE, but would be good for a kind of EXEC with parameters PREDICATE or CODE